File: internal/repository/security_transaction_repo.go
package repository

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"time"

	"github.com/zajca/zfaktury/internal/domain"
)

// SecurityTransactionRepository handles persistence of SecurityTransaction entities.
type SecurityTransactionRepository struct {
	db *sql.DB
}

// NewSecurityTransactionRepository creates a new SecurityTransactionRepository.
func NewSecurityTransactionRepository(db *sql.DB) *SecurityTransactionRepository {
	return &SecurityTransactionRepository{db: db}
}

// securityTransactionColumns is the list of columns to select for a SecurityTransaction row.
const securityTransactionColumns = `id, year, document_id, asset_type, asset_name, isin, transaction_type, transaction_date, quantity, unit_price, total_amount, fees, currency_code, exchange_rate, cost_basis, computed_gain, time_test_exempt, exempt_amount, created_at, updated_at`

// scanSecurityTransaction scans a SecurityTransaction from a row.
func scanSecurityTransaction(s scanner) (*domain.SecurityTransaction, error) {
	t := &domain.SecurityTransaction{}
	var documentID sql.NullInt64
	var isin sql.NullString
	var transactionDateStr string
	var timeTestExempt int
	var createdAtStr, updatedAtStr string

	err := s.Scan(
		&t.ID, &t.Year, &documentID, &t.AssetType, &t.AssetName, &isin,
		&t.TransactionType, &transactionDateStr, &t.Quantity,
		&t.UnitPrice, &t.TotalAmount, &t.Fees,
		&t.CurrencyCode, &t.ExchangeRate,
		&t.CostBasis, &t.ComputedGain, &timeTestExempt, &t.ExemptAmount,
		&createdAtStr, &updatedAtStr,
	)
	if err != nil {
		return nil, err
	}

	if documentID.Valid {
		t.DocumentID = &documentID.Int64
	}
	if isin.Valid {
		t.ISIN = isin.String
	}

	t.TimeTestExempt = timeTestExempt != 0

	t.TransactionDate, err = parseDate("2006-01-02", transactionDateStr)
	if err != nil {
		return nil, fmt.Errorf("scanning security_transaction transaction_date: %w", err)
	}
	t.CreatedAt, err = parseDate(time.RFC3339, createdAtStr)
	if err != nil {
		return nil, fmt.Errorf("scanning security_transaction created_at: %w", err)
	}
	t.UpdatedAt, err = parseDate(time.RFC3339, updatedAtStr)
	if err != nil {
		return nil, fmt.Errorf("scanning security_transaction updated_at: %w", err)
	}

	return t, nil
}

// Create inserts a new security transaction into the database.
func (r *SecurityTransactionRepository) Create(ctx context.Context, tx *domain.SecurityTransaction) error {
	now := time.Now()
	tx.CreatedAt = now
	tx.UpdatedAt = now

	var documentID *int64
	if tx.DocumentID != nil {
		documentID = tx.DocumentID
	}

	result, err := r.db.ExecContext(ctx, `
		INSERT INTO security_transactions (year, document_id, asset_type, asset_name, isin, transaction_type, transaction_date, quantity, unit_price, total_amount, fees, currency_code, exchange_rate, cost_basis, computed_gain, time_test_exempt, exempt_amount, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		tx.Year, documentID, tx.AssetType, tx.AssetName, tx.ISIN,
		tx.TransactionType, tx.TransactionDate.Format("2006-01-02"),
		tx.Quantity, tx.UnitPrice, tx.TotalAmount, tx.Fees,
		tx.CurrencyCode, tx.ExchangeRate,
		tx.CostBasis, tx.ComputedGain, boolToInt(tx.TimeTestExempt), tx.ExemptAmount,
		tx.CreatedAt.Format(time.RFC3339), tx.UpdatedAt.Format(time.RFC3339),
	)
	if err != nil {
		return fmt.Errorf("inserting security_transaction: %w", err)
	}

	id, err := result.LastInsertId()
	if err != nil {
		return fmt.Errorf("getting last insert id for security_transaction: %w", err)
	}
	tx.ID = id
	return nil
}

// Update modifies an existing security transaction.
func (r *SecurityTransactionRepository) Update(ctx context.Context, tx *domain.SecurityTransaction) error {
	tx.UpdatedAt = time.Now()

	var documentID *int64
	if tx.DocumentID != nil {
		documentID = tx.DocumentID
	}

	result, err := r.db.ExecContext(ctx, `
		UPDATE security_transactions SET
			year = ?, document_id = ?, asset_type = ?, asset_name = ?, isin = ?,
			transaction_type = ?, transaction_date = ?, quantity = ?,
			unit_price = ?, total_amount = ?, fees = ?,
			currency_code = ?, exchange_rate = ?,
			cost_basis = ?, computed_gain = ?, time_test_exempt = ?, exempt_amount = ?,
			updated_at = ?
		WHERE id = ?`,
		tx.Year, documentID, tx.AssetType, tx.AssetName, tx.ISIN,
		tx.TransactionType, tx.TransactionDate.Format("2006-01-02"),
		tx.Quantity, tx.UnitPrice, tx.TotalAmount, tx.Fees,
		tx.CurrencyCode, tx.ExchangeRate,
		tx.CostBasis, tx.ComputedGain, boolToInt(tx.TimeTestExempt), tx.ExemptAmount,
		tx.UpdatedAt.Format(time.RFC3339), tx.ID,
	)
	if err != nil {
		return fmt.Errorf("updating security_transaction %d: %w", tx.ID, err)
	}
	rows, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("checking rows affected for security_transaction %d: %w", tx.ID, err)
	}
	if rows == 0 {
		return fmt.Errorf("security_transaction %d: %w", tx.ID, domain.ErrNotFound)
	}
	return nil
}

// Delete removes a security transaction by ID.
func (r *SecurityTransactionRepository) Delete(ctx context.Context, id int64) error {
	result, err := r.db.ExecContext(ctx, `DELETE FROM security_transactions WHERE id = ?`, id)
	if err != nil {
		return fmt.Errorf("deleting security_transaction %d: %w", id, err)
	}

	rows, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("checking rows affected for security_transaction %d delete: %w", id, err)
	}
	if rows == 0 {
		return fmt.Errorf("security_transaction %d: %w", id, domain.ErrNotFound)
	}
	return nil
}

// GetByID retrieves a security transaction by its ID.
func (r *SecurityTransactionRepository) GetByID(ctx context.Context, id int64) (*domain.SecurityTransaction, error) {
	row := r.db.QueryRowContext(ctx, `SELECT `+securityTransactionColumns+` FROM security_transactions WHERE id = ?`, id)
	t, err := scanSecurityTransaction(row)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, fmt.Errorf("security_transaction %d: %w", id, domain.ErrNotFound)
		}
		return nil, fmt.Errorf("querying security_transaction %d: %w", id, err)
	}
	return t, nil
}

// ListByYear retrieves all security transactions for a given year, ordered by transaction_date.
func (r *SecurityTransactionRepository) ListByYear(ctx context.Context, year int) ([]domain.SecurityTransaction, error) {
	rows, err := r.db.QueryContext(ctx,
		`SELECT `+securityTransactionColumns+` FROM security_transactions WHERE year = ? ORDER BY transaction_date ASC, id ASC`,
		year,
	)
	if err != nil {
		return nil, fmt.Errorf("listing security_transactions for year %d: %w", year, err)
	}
	defer func() { _ = rows.Close() }()

	var result []domain.SecurityTransaction
	for rows.Next() {
		t, err := scanSecurityTransaction(rows)
		if err != nil {
			return nil, fmt.Errorf("scanning security_transaction row: %w", err)
		}
		result = append(result, *t)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating security_transaction rows: %w", err)
	}
	return result, nil
}

// ListByDocumentID retrieves all security transactions for a given document ID.
func (r *SecurityTransactionRepository) ListByDocumentID(ctx context.Context, documentID int64) ([]domain.SecurityTransaction, error) {
	rows, err := r.db.QueryContext(ctx,
		`SELECT `+securityTransactionColumns+` FROM security_transactions WHERE document_id = ? ORDER BY transaction_date ASC, id ASC`,
		documentID,
	)
	if err != nil {
		return nil, fmt.Errorf("listing security_transactions for document %d: %w", documentID, err)
	}
	defer func() { _ = rows.Close() }()

	var result []domain.SecurityTransaction
	for rows.Next() {
		t, err := scanSecurityTransaction(rows)
		if err != nil {
			return nil, fmt.Errorf("scanning security_transaction row: %w", err)
		}
		result = append(result, *t)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating security_transaction rows: %w", err)
	}
	return result, nil
}

// ListBuysForFIFO retrieves all buy transactions for a given asset, ordered by date for FIFO matching.
func (r *SecurityTransactionRepository) ListBuysForFIFO(ctx context.Context, assetName, assetType string) ([]domain.SecurityTransaction, error) {
	rows, err := r.db.QueryContext(ctx,
		`SELECT `+securityTransactionColumns+` FROM security_transactions WHERE asset_name = ? AND asset_type = ? AND transaction_type = 'buy' ORDER BY transaction_date ASC, id ASC`,
		assetName, assetType,
	)
	if err != nil {
		return nil, fmt.Errorf("listing buy transactions for FIFO (%s/%s): %w", assetName, assetType, err)
	}
	defer func() { _ = rows.Close() }()

	var result []domain.SecurityTransaction
	for rows.Next() {
		t, err := scanSecurityTransaction(rows)
		if err != nil {
			return nil, fmt.Errorf("scanning security_transaction row: %w", err)
		}
		result = append(result, *t)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating security_transaction rows: %w", err)
	}
	return result, nil
}

// ListSellsByYear retrieves all sell transactions for a given year, ordered by date.
func (r *SecurityTransactionRepository) ListSellsByYear(ctx context.Context, year int) ([]domain.SecurityTransaction, error) {
	rows, err := r.db.QueryContext(ctx,
		`SELECT `+securityTransactionColumns+` FROM security_transactions WHERE year = ? AND transaction_type = 'sell' ORDER BY transaction_date ASC`,
		year,
	)
	if err != nil {
		return nil, fmt.Errorf("listing sell transactions for year %d: %w", year, err)
	}
	defer func() { _ = rows.Close() }()

	var result []domain.SecurityTransaction
	for rows.Next() {
		t, err := scanSecurityTransaction(rows)
		if err != nil {
			return nil, fmt.Errorf("scanning security_transaction row: %w", err)
		}
		result = append(result, *t)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating security_transaction rows: %w", err)
	}
	return result, nil
}

// UpdateFIFOResults updates the FIFO calculation results for a security transaction.
func (r *SecurityTransactionRepository) UpdateFIFOResults(ctx context.Context, id int64, costBasis, computedGain, exemptAmount domain.Amount, timeTestExempt bool) error {
	now := time.Now()

	result, err := r.db.ExecContext(ctx, `
		UPDATE security_transactions SET cost_basis = ?, computed_gain = ?, exempt_amount = ?, time_test_exempt = ?, updated_at = ? WHERE id = ?`,
		costBasis, computedGain, exemptAmount, boolToInt(timeTestExempt), now.Format(time.RFC3339), id,
	)
	if err != nil {
		return fmt.Errorf("updating FIFO results for security_transaction %d: %w", id, err)
	}

	rows, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("checking rows affected for security_transaction %d FIFO update: %w", id, err)
	}
	if rows == 0 {
		return fmt.Errorf("security_transaction %d: %w", id, domain.ErrNotFound)
	}
	return nil
}

// DeleteByDocumentID removes all security transactions for a given document ID.
func (r *SecurityTransactionRepository) DeleteByDocumentID(ctx context.Context, documentID int64) error {
	_, err := r.db.ExecContext(ctx, `DELETE FROM security_transactions WHERE document_id = ?`, documentID)
	if err != nil {
		return fmt.Errorf("deleting security_transactions for document %d: %w", documentID, err)
	}
	return nil
}

File: rust/crates/zfaktury-db/src/repos/security_transaction_repo.rs
use crate::helpers::*;
use chrono::Local;
use rusqlite::{Connection, Row, params};
use std::sync::Mutex;
use zfaktury_core::repository::traits::SecurityTransactionRepo;
use zfaktury_domain::*;
pub struct SqliteSecurityTransactionRepo {
    conn: Mutex<Connection>,
}
impl SqliteSecurityTransactionRepo {
    pub fn new(conn: Connection) -> Self {
        Self {
            conn: Mutex::new(conn),
        }
    }
}
fn pat(s: &str) -> AssetType {
    match s {
        "stock" => AssetType::Stock,
        "etf" => AssetType::ETF,
        "bond" => AssetType::Bond,
        "fund" => AssetType::Fund,
        "crypto" => AssetType::Crypto,
        _ => AssetType::Other,
    }
}
fn ptt(s: &str) -> TransactionType {
    match s {
        "sell" => TransactionType::Sell,
        _ => TransactionType::Buy,
    }
}
fn scan(row: &Row<'_>) -> rusqlite::Result<SecurityTransaction> {
    let at: String = row.get("asset_type")?;
    let tt: String = row.get("transaction_type")?;
    let td: String = row.get("transaction_date")?;
    let c: String = row.get("created_at")?;
    let u: String = row.get("updated_at")?;
    Ok(SecurityTransaction {
        id: row.get("id")?,
        year: row.get("year")?,
        document_id: row.get("document_id")?,
        asset_type: pat(&at),
        asset_name: row.get("asset_name")?,
        isin: row.get("isin")?,
        transaction_type: ptt(&tt),
        transaction_date: parse_date_or_default(&td),
        quantity: row.get("quantity")?,
        unit_price: Amount::from_halere(row.get::<_, i64>("unit_price")?),
        total_amount: Amount::from_halere(row.get::<_, i64>("total_amount")?),
        fees: Amount::from_halere(row.get::<_, i64>("fees")?),
        currency_code: row.get("currency_code")?,
        exchange_rate: row.get("exchange_rate")?,
        cost_basis: Amount::from_halere(row.get::<_, i64>("cost_basis")?),
        computed_gain: Amount::from_halere(row.get::<_, i64>("computed_gain")?),
        time_test_exempt: row.get::<_, i32>("time_test_exempt")? != 0,
        exempt_amount: Amount::from_halere(row.get::<_, i64>("exempt_amount")?),
        created_at: parse_datetime_or_default(&c),
        updated_at: parse_datetime_or_default(&u),
    })
}
impl SecurityTransactionRepo for SqliteSecurityTransactionRepo {
    fn create(&self, tx: &mut SecurityTransaction) -> Result<(), DomainError> {
        let c = self.conn.lock().unwrap();
        let n = format_datetime(&Local::now().naive_local());
        c.execute("INSERT INTO security_transactions (year,document_id,asset_type,asset_name,isin,transaction_type,transaction_date,quantity,unit_price,total_amount,fees,currency_code,exchange_rate,cost_basis,computed_gain,time_test_exempt,exempt_amount,created_at,updated_at) VALUES (?1,?2,?3,?4,?5,?6,?7,?8,?9,?10,?11,?12,?13,?14,?15,?16,?17,?18,?18)",params![tx.year,tx.document_id,tx.asset_type.to_string(),tx.asset_name,tx.isin,tx.transaction_type.to_string(),format_date(&tx.transaction_date),tx.quantity,tx.unit_price.halere(),tx.total_amount.halere(),tx.fees.halere(),tx.currency_code,tx.exchange_rate,tx.cost_basis.halere(),tx.computed_gain.halere(),tx.time_test_exempt as i32,tx.exempt_amount.halere(),n]).map_err(|e|{log::error!("ins: {e}");DomainError::InvalidInput})?;
        tx.id = c.last_insert_rowid();
        Ok(())
    }
    fn update(&self, tx: &mut SecurityTransaction) -> Result<(), DomainError> {
        let c = self.conn.lock().unwrap();
        let n = format_datetime(&Local::now().naive_local());
        c.execute("UPDATE security_transactions SET asset_type=?1,asset_name=?2,isin=?3,transaction_type=?4,transaction_date=?5,quantity=?6,unit_price=?7,total_amount=?8,fees=?9,currency_code=?10,exchange_rate=?11,cost_basis=?12,computed_gain=?13,time_test_exempt=?14,exempt_amount=?15,updated_at=?16 WHERE id=?17",params![tx.asset_type.to_string(),tx.asset_name,tx.isin,tx.transaction_type.to_string(),format_date(&tx.transaction_date),tx.quantity,tx.unit_price.halere(),tx.total_amount.halere(),tx.fees.halere(),tx.currency_code,tx.exchange_rate,tx.cost_basis.halere(),tx.computed_gain.halere(),tx.time_test_exempt as i32,tx.exempt_amount.halere(),n,tx.id]).map_err(|e|{log::error!("upd: {e}");DomainError::InvalidInput})?;
        Ok(())
    }
    fn delete(&self, id: i64) -> Result<(), DomainError> {
        let c = self.conn.lock().unwrap();
        let r = c
            .execute("DELETE FROM security_transactions WHERE id=?1", params![id])
            .map_err(|e| {
                log::error!("del: {e}");
                DomainError::InvalidInput
            })?;
        if r == 0 {
            return Err(DomainError::NotFound);
        }
        Ok(())
    }
    fn get_by_id(&self, id: i64) -> Result<SecurityTransaction, DomainError> {
        let c = self.conn.lock().unwrap();
        c.query_row(
            "SELECT * FROM security_transactions WHERE id=?1",
            params![id],
            scan,
        )
        .map_err(|e| match e {
            rusqlite::Error::QueryReturnedNoRows => DomainError::NotFound,
            _ => {
                log::error!("q: {e}");
                DomainError::InvalidInput
            }
        })
    }
    fn list_by_year(&self, year: i32) -> Result<Vec<SecurityTransaction>, DomainError> {
        let c = self.conn.lock().unwrap();
        let mut s = c
            .prepare("SELECT * FROM security_transactions WHERE year=?1 ORDER BY transaction_date")
            .map_err(|e| {
                log::error!("p: {e}");
                DomainError::InvalidInput
            })?;
        s.query_map(params![year], scan)
            .map_err(|e| {
                log::error!("l: {e}");
                DomainError::InvalidInput
            })?
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| {
                log::error!("s: {e}");
                DomainError::InvalidInput
            })
    }
    fn list_by_document_id(
        &self,
        document_id: i64,
    ) -> Result<Vec<SecurityTransaction>, DomainError> {
        let c = self.conn.lock().unwrap();
        let mut s = c
            .prepare("SELECT * FROM security_transactions WHERE document_id=?1")
            .map_err(|e| {
                log::error!("p: {e}");
                DomainError::InvalidInput
            })?;
        s.query_map(params![document_id], scan)
            .map_err(|e| {
                log::error!("l: {e}");
                DomainError::InvalidInput
            })?
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| {
                log::error!("s: {e}");
                DomainError::InvalidInput
            })
    }
    fn list_buys_for_fifo(
        &self,
        asset_name: &str,
        asset_type: &str,
    ) -> Result<Vec<SecurityTransaction>, DomainError> {
        let c = self.conn.lock().unwrap();
        let mut s=c.prepare("SELECT * FROM security_transactions WHERE asset_name=?1 AND asset_type=?2 AND transaction_type='buy' ORDER BY transaction_date").map_err(|e|{log::error!("p: {e}");DomainError::InvalidInput})?;
        s.query_map(params![asset_name, asset_type], scan)
            .map_err(|e| {
                log::error!("l: {e}");
                DomainError::InvalidInput
            })?
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| {
                log::error!("s: {e}");
                DomainError::InvalidInput
            })
    }
    fn list_sells_by_year(&self, year: i32) -> Result<Vec<SecurityTransaction>, DomainError> {
        let c = self.conn.lock().unwrap();
        let mut s=c.prepare("SELECT * FROM security_transactions WHERE year=?1 AND transaction_type='sell' ORDER BY transaction_date").map_err(|e|{log::error!("p: {e}");DomainError::InvalidInput})?;
        s.query_map(params![year], scan)
            .map_err(|e| {
                log::error!("l: {e}");
                DomainError::InvalidInput
            })?
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| {
                log::error!("s: {e}");
                DomainError::InvalidInput
            })
    }
    fn update_fifo_results(
        &self,
        id: i64,
        cost_basis: Amount,
        computed_gain: Amount,
        exempt_amount: Amount,
        time_test_exempt: bool,
    ) -> Result<(), DomainError> {
        let c = self.conn.lock().unwrap();
        let n = format_datetime(&Local::now().naive_local());
        c.execute("UPDATE security_transactions SET cost_basis=?1,computed_gain=?2,exempt_amount=?3,time_test_exempt=?4,updated_at=?5 WHERE id=?6",params![cost_basis.halere(),computed_gain.halere(),exempt_amount.halere(),time_test_exempt as i32,n,id]).map_err(|e|{log::error!("upd fifo: {e}");DomainError::InvalidInput})?;
        Ok(())
    }
    fn delete_by_document_id(&self, document_id: i64) -> Result<(), DomainError> {
        let c = self.conn.lock().unwrap();
        c.execute(
            "DELETE FROM security_transactions WHERE document_id=?1",
            params![document_id],
        )
        .map_err(|e| {
            log::error!("del: {e}");
            DomainError::InvalidInput
        })?;
        Ok(())
    }
}
#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_db::new_test_db;
    use chrono::NaiveDate;
    #[test]
    fn test_create() {
        let c = new_test_db();
        let r = SqliteSecurityTransactionRepo::new(c);
        let mut t = SecurityTransaction {
            id: 0,
            year: 2025,
            document_id: None,
            asset_type: AssetType::Stock,
            asset_name: "AAPL".into(),
            isin: "US0378331005".into(),
            transaction_type: TransactionType::Buy,
            transaction_date: NaiveDate::from_ymd_opt(2025, 1, 15).unwrap(),
            quantity: 100,
            unit_price: Amount::from_halere(15000),
            total_amount: Amount::from_halere(1500000),
            fees: Amount::from_halere(100),
            currency_code: "USD".into(),
            exchange_rate: 10000,
            cost_basis: Amount::ZERO,
            computed_gain: Amount::ZERO,
            time_test_exempt: false,
            exempt_amount: Amount::ZERO,
            created_at: Default::default(),
            updated_at: Default::default(),
        };
        r.create(&mut t).unwrap();
        assert!(t.id > 0);
    }
}
